plugins/global-search/src/components/Results.tsx

```diff
@@ -1,108 +1,192 @@
-import { useMemo } from "react"
-import { assertNever } from "../utils/assert"
-import type { EntryResult } from "../utils/filter/group-results"
+import { defaultRangeExtractor, useVirtualizer } from "@tanstack/react-virtual"
+import { type CSSProperties, useCallback, useMemo, useRef, useState } from "react"
+import { cn } from "../utils/className"
+import type { PreparedGroup, PreparedResult } from "../utils/filter/group-results"
 import { ResultType } from "../utils/filter/types"
-import type { RootNodeType } from "../utils/indexer/types"
-import { useFocusHandlers } from "../utils/useFocus"
+import { GroupHeader } from "./GroupHeader"
 import { Match } from "./Match"
-import { IconArrowRight } from "./ui/IconArrowRight"
-import { IconCode } from "./ui/IconCode"
-import { IconCollection } from "./ui/IconCollection"
-import { IconComponent } from "./ui/IconComponent"
-import { IconWebPage } from "./ui/IconWebPage"
 
 interface ResultsProps {
-    groupedResults: readonly EntryResult[]
+    readonly groups: readonly PreparedGroup[]
 }
 
-export function ResultsList({ groupedResults }: ResultsProps) {
+export function ResultsList({ groups }: ResultsProps) {
+    const scrollElementRef = useRef<HTMLDivElement>(null)
+    const activeStickyIndexRef = useRef(0)
+    const { collapsedGroups, toggleGroup } = useCollapsedGroups()
+    const { virtualItems, stickyIndexes } = useProcessedResults(groups, collapsedGroups)
+
+    const isActiveSticky = useCallback((index: number) => activeStickyIndexRef.current === index, [])
+
+    const rowVirtualizer = useVirtualizer({
+        overscan: 10,
+        count: virtualItems.length,
+        getScrollElement: () => scrollElementRef.current,
+        estimateSize: index => {
+            const item = virtualItems[index]
+            if (!item) return 0
+            if (item.type === "group-header") return 35
+
+            const nextItem = virtualItems[index + 1]
+            const isLastMatchInGroup = nextItem?.groupId !== item.groupId
+            return isLastMatchInGroup ? 35 : 30
+        },
+        rangeExtractor: range => {
+            activeStickyIndexRef.current = stickyIndexes.findLast(index => range.startIndex >= index) ?? 0
+            const next = new Set([activeStickyIndexRef.current, ...defaultRangeExtractor(range)])
+            return [...next].sort((a, b) => a - b)
```

Comment:
This also feels potentially expensive. `defaultRangeExtractor(range)` already returns a sorted list of numbers. Inserting one (if not present) should be achievable with a single pass + splice.

Reply:
same as above. Additionally here: this contains (depending on the screen size) maybe 50 items. Sorting them like is is most likely fine

Reply:
Ok! Sorting - sure, especially if it's pre-mostly-sorted, but it also instantiates a `Set` 🥲

Reply:
Will cover this when also looking into the thing above. Adding a TODO comment here.